File: app/src/main/java/im/tox/antox/utils/CaptureAudio.java
package im.tox.antox.utils;

import android.media.AudioFormat;
import android.media.AudioRecord;
import android.media.MediaRecorder;
import android.os.AsyncTask;
import android.util.Log;

import im.tox.antox.tox.ToxSingleton;

/**
 * Created by soft on 18/09/14.
 */
public class CaptureAudio extends AsyncTask<String, Void, Void> {
    private int bufferSizeBytes;

    @Override
    protected Void doInBackground(String... params) {
        ToxSingleton toxSingleton = ToxSingleton.getInstance();
        AudioRecord audioRecord = findAudioRecord();
        audioRecord.startRecording();

        while(true) {
            if(isCancelled())
                break;

            try {
                byte[] buffer = new byte[bufferSizeBytes];
                int read = audioRecord.read(buffer, 0, bufferSizeBytes);
                toxSingleton.jTox.avSendAudio(Integer.parseInt(params[0]), buffer);
                Log.d("Mic", "Sending audio to:" + params[0]);
            } catch (Exception e) {
            }
        }

        audioRecord.stop();
        audioRecord.release();

        return null;
    }

    private static int[] mSampleRates = new int[] { 8000, 11025, 22050, 44100 };

    private AudioRecord findAudioRecord() {
        for (int rate : mSampleRates) {
            for (short audioFormat : new short[] { AudioFormat.ENCODING_PCM_8BIT, AudioFormat.ENCODING_PCM_16BIT }) {
                for (short channelConfig : new short[] { AudioFormat.CHANNEL_IN_MONO, AudioFormat.CHANNEL_IN_STEREO }) {
                    try {
                        Log.d("CaptureAudio", "Attempting rate " + rate + "Hz, bits: " + audioFormat + ", channel: "
                                + channelConfig);
                        int bufferSize = AudioRecord.getMinBufferSize(rate, channelConfig, audioFormat);

                        if (bufferSize != AudioRecord.ERROR_BAD_VALUE) {
                            // check if we can instantiate and have a success
                            AudioRecord recorder = new AudioRecord(MediaRecorder.AudioSource.DEFAULT, rate, channelConfig, audioFormat, bufferSize);

                            if (recorder.getState() == AudioRecord.STATE_INITIALIZED) {
                                bufferSizeBytes = bufferSize;
                                return recorder;
                            }

                        }
                    } catch (Exception e) {
                    }
                }
            }
        }
        return null;
    }
}

File: app/src/main/scala/im/tox/antox/utils/CaptureAudio.scala
package im.tox.antox.utils

import android.media.{MediaRecorder, AudioFormat, AudioRecord}
import android.os.AsyncTask
import android.util.Log
import im.tox.antox.tox.ToxSingleton
import im.tox.jtoxcore.ToxCodecSettings
import rx.lang.scala.Observable

object CaptureAudio {

  val TAG = "im.tox.antox.utils.CaptureAudio"

  var bufferSizeBytes: Int = _
  private val sampleRates: Array[Int] = Array(8000, 11025, 22050, 44100)

  def makeObservable(callID: Integer, codecSettings: ToxCodecSettings): Observable[Boolean] = {
    Observable[Boolean](subscriber => {
      val mAudioRecord = findAudioRecord()
      val rtpPayloadSize = 65535
      mAudioRecord match {
        case Some(audioRecord) => {
          audioRecord.startRecording()

          while (!subscriber.isUnsubscribed) {
            try {
              val buffer = Array.ofDim[Short](CaptureAudio.bufferSizeBytes)
              audioRecord.read(buffer, 0, CaptureAudio.bufferSizeBytes)
              val intBuffer = buffer.map(x => x: Int)
              val preparedBuffer = ToxSingleton.jTox.avPrepareAudioFrame(callID,
                rtpPayloadSize, intBuffer, codecSettings.audio_frame_duration)
              ToxSingleton.jTox.avSendAudio(callID, preparedBuffer)
              Log.d("Mic", "Sending audio to:" + callID)
            } catch {
              case e: Exception => 
                e.printStackTrace
                subscriber.onError(e)
            }
          }

          audioRecord.stop()
          audioRecord.release()
        }
        case None => Log.d(TAG, "Audio record: None!")
      }
      subscriber.onCompleted()
    })
  }

  private def findAudioRecord(): Option[AudioRecord] = {
    try {
      val audioFormat = AudioFormat.ENCODING_PCM_16BIT
      val channelConfig = AudioFormat.CHANNEL_IN_MONO
      val rate = 48000
      Log.d("CaptureAudio", "Attempting rate " + rate + "Hz, bits: " + audioFormat +
        ", channel: " +
        channelConfig)
      val bufferSize = AudioRecord.getMinBufferSize(rate, channelConfig, audioFormat)
      if (bufferSize != AudioRecord.ERROR_BAD_VALUE) {
        val recorder = new AudioRecord(0, rate, channelConfig, audioFormat,
          bufferSize)
        if (recorder.getState == AudioRecord.STATE_INITIALIZED) {
          CaptureAudio.bufferSizeBytes = bufferSize
          return Some(recorder)
        } else {
          None
        }
      } else {
        None
      }
    } catch {
      case e: Exception => 
        e.printStackTrace
        None
    }
  }
}
